File: app/src/main/java/inha/nslab/easytrack/GoogleAuthActivity.java
package inha.nslab.easytrack;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.SignInButton;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;

import java.util.Locale;
import java.util.concurrent.TimeUnit;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.StatusRuntimeException;

public class GoogleAuthActivity extends AppCompatActivity {
    private static final int RC_SIGN_IN_WITH_GOOGLE = 101;
    // private static String TAG = "ET_AUTH_EXAMPLE_APP";
    private GoogleSignInClient signInClient;
    private SignInButton signInButton;
    private SignInButton continueWithGoogleAccountButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_google_auth);

        signInButton = findViewById(R.id.googleSignInButton);
        continueWithGoogleAccountButton = findViewById(R.id.continueWithGoogleAccountButton);

        // Google login client setup
        GoogleSignInOptions googleSignInOptions = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.server_google_client_id))
                .requestEmail()
                .build();
        signInClient = GoogleSignIn.getClient(this, googleSignInOptions);

        signInButton.setOnClickListener(view -> {
            continueWithGoogleAccountButton.setVisibility(View.GONE);
            signInButton.setEnabled(false);
            GoogleSignInAccount account = GoogleSignIn.getLastSignedInAccount(this);
            if (account == null)
                startGoogleAuthenticationActivity();
            else
                signInClient.signOut().addOnSuccessListener(v -> {
                    continueWithGoogleAccountButton.setVisibility(View.GONE);
                    setGoogleSignInButtonText(signInButton, "Sign in");
                    startGoogleAuthenticationActivity();
                }).addOnFailureListener(e -> {
                    GoogleSignInAccount _account = GoogleSignIn.getLastSignedInAccount(this);
                    assert _account != null;
                    continueWithGoogleAccountButton.setVisibility(View.VISIBLE);
                    signInButton.setEnabled(true);
                    setGoogleSignInButtonText(continueWithGoogleAccountButton, "Continue as: " + _account.getEmail());
                    setGoogleSignInButtonText(signInButton, "Sign in with different account");
                });
        });
        continueWithGoogleAccountButton.setOnClickListener(view -> {
            GoogleSignInAccount account = GoogleSignIn.getLastSignedInAccount(this);
            assert account != null;
            continueWithGoogleAccountButton.setEnabled(false);
            signInButton.setEnabled(false);
            new Thread(() -> {
                try {
                    ManagedChannel channel = ManagedChannelBuilder.forAddress(
                            getString(R.string.grpc_host),
                            Integer.parseInt(getString(R.string.grpc_port))
                    ).usePlaintext().build();

                    ETServiceGrpc.ETServiceBlockingStub stub = ETServiceGrpc.newBlockingStub(channel);
                    EtService.LoginWithGoogleIdTokenRequestMessage requestMessage = EtService.LoginWithGoogleIdTokenRequestMessage.newBuilder()
                            .setIdToken(account.getIdToken())
                            .build();
                    EtService.LoginResponseMessage responseMessage = stub.loginWithGoogleId(requestMessage);

                    try {
                        channel.shutdown().awaitTermination(1, TimeUnit.SECONDS);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }

                    if (responseMessage.getDoneSuccessfully())
                        runOnUiThread(() -> {
                            Intent result = new Intent("etAuthResult");
                            result.putExtra("fields", "idToken,fullName,email,userId");
                            result.putExtra("idToken", account.getIdToken());
                            result.putExtra("fullName", account.getEmail());
                            result.putExtra("email", account.getEmail());
                            result.putExtra("userId", responseMessage.getUserId());
                            setResult(Activity.RESULT_OK, result);
                            finish();
                        });
                    else
                        // technical issue, shouldn't happen
                        runOnUiThread(() -> signInClient.signOut().addOnCompleteListener((Void) -> {
                            Intent result = new Intent("etAuthResult");
                            result.putExtra("fields", "note");
                            result.putExtra("note", String.format(Locale.getDefault(), "please contact the EasyTrack developers with the following details: success=%b, userId=%d", responseMessage.getDoneSuccessfully(), responseMessage.getUserId()));
                            setResult(Activity.RESULT_CANCELED, result);
                            finish();
                        }));
                } catch (StatusRuntimeException e) {
                    runOnUiThread(() -> {
                        Intent result = new Intent("etAuthResult");
                        result.putExtra("fields", "exception_message,exception_details");
                        result.putExtra("exception_message", e.getMessage());
                        result.putExtra("exception_details", e.toString());
                        setResult(Activity.RESULT_CANCELED, result);
                        finish();
                    });
                }
            }).start();
        });

        GoogleSignInAccount account = GoogleSignIn.getLastSignedInAccount(this);
        if (account == null) {
            setGoogleSignInButtonText(signInButton, "Sign in");
            continueWithGoogleAccountButton.setVisibility(View.GONE);
            startGoogleAuthenticationActivity();
        } else {
            setGoogleSignInButtonText(signInButton, "Sign in with different account");
            continueWithGoogleAccountButton.setVisibility(View.VISIBLE);
            setGoogleSignInButtonText(continueWithGoogleAccountButton, "Continue with: " + account.getEmail());
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == RC_SIGN_IN_WITH_GOOGLE) {
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);
            try {
                GoogleSignInAccount account = task.getResult(ApiException.class);
                if (account != null) {
                    new Thread(() -> {
                        try {
                            ManagedChannel channel = ManagedChannelBuilder.forAddress(
                                    getString(R.string.grpc_host),
                                    Integer.parseInt(getString(R.string.grpc_port))
                            ).usePlaintext().build();

                            ETServiceGrpc.ETServiceBlockingStub stub = ETServiceGrpc.newBlockingStub(channel);
                            EtService.LoginWithGoogleIdTokenRequestMessage requestMessage = EtService.LoginWithGoogleIdTokenRequestMessage.newBuilder()
                                    .setIdToken(account.getIdToken())
                                    .build();
                            EtService.LoginResponseMessage responseMessage = stub.loginWithGoogleId(requestMessage);

                            try {
                                channel.shutdown().awaitTermination(1, TimeUnit.SECONDS);
                            } catch (InterruptedException e) {
                                e.printStackTrace();
                            }

                            if (responseMessage.getDoneSuccessfully())
                                runOnUiThread(() -> {
                                    Intent result = new Intent("etAuthResult");
                                    result.putExtra("fields", "idToken,fullName,email,userId");
                                    result.putExtra("idToken", account.getIdToken());
                                    result.putExtra("fullName", account.getEmail());
                                    result.putExtra("email", account.getEmail());
                                    result.putExtra("userId", responseMessage.getUserId());
                                    setResult(Activity.RESULT_OK, result);
                                    finish();
                                });
                            else
                                // technical issue, shouldn't happen
                                runOnUiThread(() -> signInClient.signOut().addOnCompleteListener((Void) -> {
                                    Intent result = new Intent("etAuthResult");
                                    result.putExtra("fields", "note");
                                    result.putExtra("note", String.format(Locale.getDefault(), "please contact the EasyTrack developers with the following details: success=%b, userId=%d", responseMessage.getDoneSuccessfully(), responseMessage.getUserId()));
                                    setResult(Activity.RESULT_CANCELED, result);
                                    finish();
                                }));
                        } catch (StatusRuntimeException e) {
                            runOnUiThread(() -> {
                                Intent result = new Intent("etAuthResult");
                                result.putExtra("fields", "exception_message,exception_details");
                                result.putExtra("exception_message", e.getMessage());
                                result.putExtra("exception_details", e.toString());
                                setResult(Activity.RESULT_CANCELED, result);
                                finish();
                            });
                        }
                    }).start();
                } else {
                    Intent result = new Intent("etAuthResult");
                    result.putExtra("fields", "N/A");
                    setResult(Activity.RESULT_FIRST_USER, result);
                    finish();
                }
            } catch (ApiException e) {
                Intent result = new Intent("etAuthResult");
                result.putExtra("fields", "exception_message,exception_details");
                result.putExtra("exception_message", e.getMessage());
                result.putExtra("exception_details", e.toString());
                setResult(Activity.RESULT_CANCELED, result);
                finish();
            }
        }
    }

    private void setGoogleSignInButtonText(SignInButton button, String text) {
        for (int n = 0; n < button.getChildCount(); n++) {
            View view = button.getChildAt(n);
            if (view instanceof TextView) {
                ((TextView) view).setText(text);
                return;
            }
        }
    }

    private void startGoogleAuthenticationActivity() {
        Intent intent = signInClient.getSignInIntent();
        intent.setFlags(0);
        startActivityForResult(intent, RC_SIGN_IN_WITH_GOOGLE);
    }
}

File: app/src/main/java/inha/nsl/easytrack/GoogleAuthActivity.kt
package inha.nsl.easytrack

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.children
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.SignInButton
import com.google.android.gms.common.api.ApiException
import io.grpc.ManagedChannelBuilder
import io.grpc.StatusRuntimeException
import kotlinx.android.synthetic.main.activity_google_auth.*
import java.util.*
import java.util.concurrent.TimeUnit

class GoogleAuthActivity : AppCompatActivity() {
    private lateinit var signInClient: GoogleSignInClient
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_google_auth)

        // Google login client setup
        val googleSignInOptions = GoogleSignInOptions.Builder()
            .requestProfile()
            .requestEmail()
            .requestIdToken(getString(R.string.default_web_client_id))
            .build()
        signInClient = GoogleSignIn.getClient(this, googleSignInOptions)
        googleSignInButton.setOnClickListener {
            continueWithGoogleAccountButton.visibility = View.GONE
            googleSignInButton.isEnabled = false
            val account = GoogleSignIn.getLastSignedInAccount(this)
            if (account == null || account.isExpired)
                startGoogleAuthenticationActivity()
            else {
                signInClient.signOut().addOnSuccessListener {
                    continueWithGoogleAccountButton.visibility = View.GONE
                    googleSignInButton.isEnabled = true
                    setGoogleSignInButtonText(googleSignInButton, "Sign in")
                    startGoogleAuthenticationActivity()
                }.addOnFailureListener {
                    val acc = GoogleSignIn.getLastSignedInAccount(this)!!
                    continueWithGoogleAccountButton.visibility = View.VISIBLE
                    googleSignInButton.isEnabled = true
                    setGoogleSignInButtonText(
                        continueWithGoogleAccountButton,
                        "Continue as: " + acc.email
                    )
                    setGoogleSignInButtonText(googleSignInButton, "Sign in with different account")
                }
            }
        }
        continueWithGoogleAccountButton.setOnClickListener {
            val account = GoogleSignIn.getLastSignedInAccount(this)!!
            continueWithGoogleAccountButton.isEnabled = false
            googleSignInButton.isEnabled = false
            val thread = Thread {
                try {
                    val channel = ManagedChannelBuilder.forAddress(
                        getString(R.string.grpc_host),
                        getString(R.string.grpc_port).toInt()
                    ).usePlaintext().build()
                    val stub = ETServiceGrpc.newBlockingStub(channel)
                    val requestMessage =
                        EtService.LoginWithGoogle.Request.newBuilder().setIdToken(account.idToken)
                            .build()
                    val responseMessage = stub.loginWithGoogle(requestMessage)
                    try {
                        channel.shutdown().awaitTermination(1, TimeUnit.SECONDS)
                    } catch (e: InterruptedException) {
                        e.printStackTrace()
                    }
                    if (responseMessage.success) runOnUiThread {
                        val result = Intent("etAuthResult")
                        result.putExtra("resultFieldNames", "name,email,userId,sessionKey")
                        result.putExtra("name", account.displayName)
                        result.putExtra("email", account.email)
                        result.putExtra("userId", responseMessage.userId)
                        result.putExtra("sessionKey", responseMessage.sessionKey)
                        setResult(Activity.RESULT_OK, result)
                        finish()
                    } else  // technical issue, shouldn't happen
                        runOnUiThread {
                            signInClient.signOut().addOnCompleteListener {
                                val result = Intent("etAuthResult")
                                result.putExtra("resultFieldNames", "note")
                                result.putExtra(
                                    "note",
                                    String.format(
                                        Locale.getDefault(),
                                        "please contact the EasyTrack developers with the following details: success=%b, sessionKey=%d",
                                        responseMessage.success,
                                        responseMessage.sessionKey
                                    )
                                )
                                setResult(Activity.RESULT_CANCELED, result)
                                finish()
                            }
                        }
                } catch (e: StatusRuntimeException) {
                    runOnUiThread {
                        val result = Intent("etAuthResult")
                        result.putExtra("resultFieldNames", "exception_message,exception_details")
                        result.putExtra("exception_message", e.message)
                        result.putExtra("exception_details", e.toString())
                        setResult(Activity.RESULT_CANCELED, result)
                        finish()
                    }
                }
            }
            thread.start()
            thread.join()
        }
        val account = GoogleSignIn.getLastSignedInAccount(this)
        if (account == null || account.isExpired) {
            setGoogleSignInButtonText(googleSignInButton, "Sign in")
            continueWithGoogleAccountButton.visibility = View.GONE
            startGoogleAuthenticationActivity()
        } else {
            setGoogleSignInButtonText(googleSignInButton, "Sign in with different account")
            continueWithGoogleAccountButton.visibility = View.VISIBLE
            setGoogleSignInButtonText(
                continueWithGoogleAccountButton,
                "Continue as: ${account.email}"
            )
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == RC_SIGN_IN_WITH_GOOGLE) {
            val task = GoogleSignIn.getSignedInAccountFromIntent(data)
            try {
                val account = task.getResult(ApiException::class.java)
                if (account != null) {
                    Thread {
                        try {
                            val channel = ManagedChannelBuilder.forAddress(
                                getString(R.string.grpc_host),
                                getString(R.string.grpc_port).toInt()
                            ).usePlaintext().build()
                            val stub = ETServiceGrpc.newBlockingStub(channel)
                            val requestMessage = EtService.LoginWithGoogle.Request.newBuilder()
                                .setIdToken(account.idToken).build()
                            val responseMessage = stub.loginWithGoogle(requestMessage)
                            channel.shutdown()
                            if (responseMessage.success) runOnUiThread {
                                val result = Intent("etAuthResult")
                                result.putExtra("resultFieldNames", "name,email,userId,sessionKey")
                                result.putExtra("name", account.displayName)
                                result.putExtra("email", account.email)
                                result.putExtra("userId", responseMessage.userId)
                                result.putExtra("sessionKey", responseMessage.sessionKey)
                                setResult(Activity.RESULT_OK, result)
                                finish()
                            } else  // technical issue, shouldn't happen
                                runOnUiThread {
                                    signInClient.signOut().addOnCompleteListener {
                                        val result = Intent("etAuthResult")
                                        result.putExtra("resultFieldNames", "note")
                                        result.putExtra(
                                            "note",
                                            String.format(
                                                Locale.getDefault(),
                                                "please contact the EasyTrack developers with the following details: success=%b, sessionKey=%d",
                                                responseMessage.success,
                                                responseMessage.sessionKey
                                            )
                                        )
                                        setResult(Activity.RESULT_CANCELED, result)
                                        finish()
                                    }
                                }
                        } catch (e: StatusRuntimeException) {
                            runOnUiThread {
                                val result = Intent("etAuthResult")
                                result.putExtra(
                                    "resultFieldNames",
                                    "exception_message,exception_details"
                                )
                                result.putExtra("exception_message", e.message)
                                result.putExtra("exception_details", e.toString())
                                setResult(Activity.RESULT_CANCELED, result)
                                finish()
                            }
                        }
                    }.start()
                } else {
                    val result = Intent("etAuthResult")
                    result.putExtra("resultFieldNames", "N/A")
                    setResult(Activity.RESULT_FIRST_USER, result)
                    finish()
                }
            } catch (e: ApiException) {
                val result = Intent("etAuthResult")
                result.putExtra("resultFieldNames", "exception_message,exception_details")
                result.putExtra("exception_message", e.message)
                result.putExtra("exception_details", e.toString())
                setResult(Activity.RESULT_CANCELED, result)
                finish()
            }
        }
    }

    private fun setGoogleSignInButtonText(button: SignInButton, text: String) {
        for (child in button.children) {
            if (child is TextView) {
                child.text = text
                return
            }
        }
    }

    private fun startGoogleAuthenticationActivity() {
        val intent = signInClient.signInIntent
        intent.flags = 0
        startActivityForResult(intent, RC_SIGN_IN_WITH_GOOGLE)
    }

    companion object {
        private const val RC_SIGN_IN_WITH_GOOGLE = 101
    }
}
